File: algorithms/sorting/quick sort/quick_sort.py
def quick_sort(arr):
    if len(arr) <= 1:
        return arr

    # pick a pivot point
    pivot = arr[0]
    # separate all values smaller or larger than pivot into separate lists
    smaller = []
    larger = []

    for i in range(1, len(arr)):
        if arr[i] <= pivot:
            smaller.append(arr[i])
        else:
            larger.append(arr[i])

    # sort smaller and larger
    smaller = quick_sort(smaller)
    larger = quick_sort(larger)
    # concat smaller + pivot + larger
    return smaller + [pivot] + larger


values = [9, 8, 7, 6, 5, 4, 3, 2, 1]
result = quick_sort(values)
print(result)

File: algorithms/sorting/quick sort/quickSort.js
function quickSort(arr) {
  // base case
  if (arr.length <= 1) {
    return arr;
  }
  // pick a pivot
  const pivot = arr[0];
  // separate all items smaller or larger than pivot into separate arrays
  let smaller = [];
  let larger = [];

  // loop over all elems but skip the pivot (first element in our case)
  for (let i = 1; i < arr.length; i++) {
    if (arr[i] <= pivot) {
      smaller.push(arr[i]);
    } else {
      larger.push(arr[i]);
    }
  }
  // sort smaller and larger arrays in place
  smaller = quickSort(smaller);
  larger = quickSort(larger);

  // return concat of smaller + pivot + larger
  return [...smaller, pivot, ...larger];
}

const values = [9, 8, 7, 6, 5, 4, 3, 2, 1];

const result = quickSort(values);
console.log(result);
